### PR TITLE
Fix product position in empty category

### LINKED ISSUE #14903: Cannot add a position_in_category using webservices
###Describe the bug
I´m trying to change the position of a product into category using the prestashop api but don´t working for me since that webservice generates a ramdon position skiping that i have done.

To reproduce this error I use the postman tool and the xml product structure

### Expected behavior

If we have an empty category and we set position_in_category to 1, the product is added into the category.

### Steps to reproduce

1 . have empty category
2 . call API to create a product into that category passing position_in_category equals to 1.

#### Acceptance Test
Unskip tests 
* https://github.com/PrestaShop/PrestaShop/blob/d6bd4586da1d519e3ccb6ce2dd2420b9a755b5b2/tests/UI/campaigns/functional/WS/02_productsCRUD.ts#L442-L443
* https://github.com/PrestaShop/PrestaShop/blob/d6bd4586da1d519e3ccb6ce2dd2420b9a755b5b2/tests/UI/campaigns/functional/WS/02_productsCRUD.ts#L1123-L1124

### Additional information
PrestaShop version: 1.7.5.0 & 1.7.8.8
PHP version: 7.0.33

### LINKED ISSUE #26761: SetWsPositionInCategory (Product.php#L7117) does not support add first product in an empty category
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

Looking at code, we think that the case where the category is empty (not contain products) and we would like to add a product to that category in the first position ($postion = 1) does not works because the query to retrieve number of products into category return 0 and the check
if ($position > count($result))  is true and goes in error with 'You cannot set a position greater than the total number of products in the category, starting at 1.' message.
 the 

### Expected behavior

If we have an empty category and we set position_in_category to 1, the product is added into the category.

### Steps to reproduce

1 . have empty category
2 . call API to create a product into that category passing position_in_category equals to 1.

### PrestaShop version(s) where the bug happened

1.7.8

### PHP version(s) where the bug happened

_No response_

### If your bug is related to a module, specify its name and its version

_No response_

### LINKED ISSUE #31340: Not able to set position of a product in an empty category
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

I'm trying to push a product to a new created category that is still empty using the WS.  
My product have these keys and values :  
`id_category_default` => `my_new_id`  
`position_in_category` => `1`.

I'm getting a 500 error (code 135). because the position is higher than the total number of products in the category (0 currently).

### Expected behavior

It should accept new products with position being (current product count + 1).

### Steps to reproduce

Try to push a product in a empty category in `id_category_default` with `position_in_category` set to 1. It should work properly. Same as if you want to set a new category to a product. The category containing 3 products you want your product to be in 4th place. So push the `id_category_default` and `position_in_category` set to 4. It should work. Then test again the same request but with `position_in_category` set to 5, it should not work.

### PrestaShop version(s) where the bug happened

1.7.8.8 and develop

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by TomLorenzi**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop / 1.7.8.8
| Description?      | When trying to push a product with the key position_in_category to 1 in an empty category you will trigger the error 135, saying that the product position is higher than the category's products count. This sould normally work. It should also work if I have already 3 products in my category and I want to set a new one to the 4th position.
| Type?             | bug fix / improvement
| Category?         | WS
| BC breaks?        | ?
| Deprecations?     | no
| How to test?      | Try to push a product in a empty category in `id_category_default` with `position_in_category` set to 1. It should work properly. Same as if you want to set a new category to a product. The category containing 3 products you want your product to be in 4th place. So push the `id_category_default` and `position_in_category` set to 4. It should work. Then test again the same request but with `position_in_category` set to 5, it should not work.
| Fixed ticket?     | Fixes #31340 , Fixes #14903 , Fixes #26761
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

**Comment by rodriciru** [comment]
It works!

**Comment by kpodemski** [comment]
Hi @TomLorenzi 

There are some issues with your PR reported by our CI :(

- Coding standards
- You have a merge commit inside your commits - it has to be a clean git history, this is why it is required to use `git pull` with a rebase

**Comment by TomLorenzi** [comment]
> Hi @TomLorenzi
> 
> There are some issues with your PR reported by our CI :(
> 
>     * Coding standards
> 
>     * You have a merge commit inside your commits - it has to be a clean git history, this is why it is required to use `git pull` with a rebase

I will rebase and fix the commits, but where are the coding standards written ? I can't find them :(

**Comment by TomLorenzi** [comment]
@kpodemski commit history has been clean, you can now remove the waiting for author label and waiting for rebase label :)

File: classes/Product.php [hunk 1]
```diff
@@ -7159,7 +7159,11 @@ public function setWsPositionInCategory($position)
             'ORDER BY `position`'
         );
 
-        if ($position > count($result)) {
+        $listIds = array_column($result, 'id_product');
+        $count = count($result);
+        if ($position > $count && in_array($this->id, $listIds)
+            || $position > $count + 1 && !in_array($this->id, $listIds)
```
Comment (kpodemski) [inline classes/Product.php]:
```suggestion
            || $position > $count + 1 && !in_array($this->id, $listIds)
```

**Comment by TomLorenzi** [comment]
Thanks for the cs fixer, my bad when i see what it was :/

**Review by kpodemski — APPROVED** [review]

**Review by nicosomb — APPROVED** [review]

**Review by hibatallahAouadni — COMMENTED** [review]
Hello @TomLorenzi 

Thanks for your PR :rocket: 
Your PR fixed the issue about `position_in_category` :tada: 
I use this body request:
```
<prestashop>
<product>
	<id_tax_rules_group >1</id_tax_rules_group>
	<id_shop_default>1</id_shop_default>
    <associations>
    <categories>
      <category><id>10</id></category>
      <position_in_category>1</position_in_category>
    </categories>
    </associations>
	<reference>chocotom</reference>
  	<price>25</price>
	<state>1</state>
	<minimal_quantity>1</minimal_quantity>
	<active>1</active>
  	<id_manufacturer/>
  	<id_supplier>1</id_supplier>
  	<link_rewrite>pack</link_rewrite>
	<available_for_order>1</available_for_order>
	<condition>new</condition>
	<show_price>1</show_price>
	<visibility>both</visibility>
	<name><language id="1" >ChocoTom</language></name>
	<description><language id="1">New ChocoTom</language></description>
	<description_short><language id="1" >New ChocoTom short</language></description_short>
</product>
</prestashop>
```
And I get this as result:
```
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
    <product>
        <id>
            <![CDATA[25]]>
        </id>
        <id_manufacturer>
            <![CDATA[]]>
        </id_manufacturer>
        <id_supplier xlink:href="http://localhost:8081/develop/PrestaShop/api/suppliers/1">
            <![CDATA[1]]>
        </id_supplier>
        <id_category_default>
            <![CDATA[]]>
        </id_category_default>
        <new>
            <![CDATA[]]>
        </new>
        <cache_default_attribute>
            <![CDATA[]]>
        </cache_default_attribute>
        <id_default_image notFilterable="true">
            <![CDATA[]]>
        </id_default_image>
        <id_default_combination notFilterable="true">
            <![CDATA[0]]>
        </id_default_combination>
        <id_tax_rules_group xlink:href="http://localhost:8081/develop/PrestaShop/api/tax_rule_groups/1">
            <![CDATA[1]]>
        </id_tax_rules_group>
        <position_in_category notFilterable="true">
            <![CDATA[2]]>
        </position_in_category>
        <manufacturer_name notFilterable="true">
            <![CDATA[]]>
        </manufacturer_name>
        <quantity notFilterable="true">
            <![CDATA[]]>
        </quantity>
        <type notFilterable="true">
            <![CDATA[simple]]>
        </type>
        <id_shop_default>
            <![CDATA[1]]>
        </id_shop_default>
        <reference>
            <![CDATA[chocotom]]>
        </reference>
        <supplier_reference>
            <![CDATA[]]>
        </supplier_reference>
        <location>
            <![CDATA[]]>
        </location>
        <width>
            <![CDATA[]]>
        </width>
        <height>
            <![CDATA[]]>
        </height>
        <depth>
            <![CDATA[]]>
        </depth>
        <weight>
            <![CDATA[]]>
        </weight>
        <quantity_discount>
            <![CDATA[]]>
        </quantity_discount>
        <ean13>
            <![CDATA[]]>
        </ean13>
        <isbn>
            <![CDATA[]]>
        </isbn>
        <upc>
            <![CDATA[]]>
        </upc>
        <mpn>
            <![CDATA[]]>
        </mpn>
        <cache_is_pack>
            <![CDATA[]]>
        </cache_is_pack>
        <cache_has_attachments>
            <![CDATA[]]>
        </cache_has_attachments>
        <is_virtual>
            <![CDATA[]]>
        </is_virtual>
        <state>
            <![CDATA[1]]>
        </state>
        <additional_delivery_times>
            <![CDATA[]]>
        </additional_delivery_times>
        <delivery_in_stock>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </delivery_in_stock>
        <delivery_out_stock>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </delivery_out_stock>
        <product_type>
            <![CDATA[]]>
        </product_type>
        <on_sale>
            <![CDATA[]]>
        </on_sale>
        <online_only>
            <![CDATA[]]>
        </online_only>
        <ecotax>
            <![CDATA[]]>
        </ecotax>
        <minimal_quantity>
            <![CDATA[1]]>
        </minimal_quantity>
        <low_stock_threshold>
            <![CDATA[]]>
        </low_stock_threshold>
        <low_stock_alert>
            <![CDATA[]]>
        </low_stock_alert>
        <price>
            <![CDATA[25]]>
        </price>
        <wholesale_price>
            <![CDATA[]]>
        </wholesale_price>
        <unity>
            <![CDATA[]]>
        </unity>
        <unit_price>
            <![CDATA[]]>
        </unit_price>
        <unit_price_ratio>
            <![CDATA[]]>
        </unit_price_ratio>
        <additional_shipping_cost>
            <![CDATA[]]>
        </additional_shipping_cost>
        <customizable>
            <![CDATA[]]>
        </customizable>
        <text_fields>
            <![CDATA[]]>
        </text_fields>
        <uploadable_files>
            <![CDATA[]]>
        </uploadable_files>
        <active>
            <![CDATA[1]]>
        </active>
        <redirect_type>
            <![CDATA[]]>
        </redirect_type>
        <id_type_redirected>
            <![CDATA[]]>
        </id_type_redirected>
        <available_for_order>
            <![CDATA[1]]>
        </available_for_order>
        <available_date>
            <![CDATA[]]>
        </available_date>
        <show_condition>
            <![CDATA[]]>
        </show_condition>
        <condition>
            <![CDATA[new]]>
        </condition>
        <show_price>
            <![CDATA[1]]>
        </show_price>
        <indexed>
            <![CDATA[]]>
        </indexed>
        <visibility>
            <![CDATA[both]]>
        </visibility>
        <advanced_stock_management>
            <![CDATA[]]>
        </advanced_stock_management>
        <date_add>
            <![CDATA[2023-06-07 10:47:00]]>
        </date_add>
        <date_upd>
            <![CDATA[2023-06-07 10:47:00]]>
        </date_upd>
        <pack_stock_type>
            <![CDATA[]]>
        </pack_stock_type>
        <meta_description>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </meta_description>
        <meta_keywords>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </meta_keywords>
        <meta_title>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </meta_title>
        <link_rewrite>
            <language id="1">
                <![CDATA[a]]>
            </language>
            <language id="2">
                <![CDATA[c]]>
            </language>
        </link_rewrite>
        <name>
            <language id="1" xlink:href="http://localhost:8081/develop/PrestaShop/api/languages/1">
                <![CDATA[ChocoTom]]>
            </language>
            <language id="2" xlink:href="http://localhost:8081/develop/PrestaShop/api/languages/2">
                <![CDATA[]]>
            </language>
        </name>
        <description>
            <language id="1" xlink:href="http://localhost:8081/develop/PrestaShop/api/languages/1">
                <![CDATA[New ChocoTom]]>
            </language>
            <language id="2" xlink:href="http://localhost:8081/develop/PrestaShop/api/languages/2">
                <![CDATA[]]>
            </language>
        </description>
        <description_short>
            <language id="1" xlink:href="http://localhost:8081/develop/PrestaShop/api/languages/1">
                <![CDATA[New ChocoTom short]]>
            </language>
            <language id="2" xlink:href="http://localhost:8081/develop/PrestaShop/api/languages/2">
                <![CDATA[]]>
            </language>
        </description_short>
        <available_now>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </available_now>
        <available_later>
            <language id="1">
                <![CDATA[]]>
            </language>
            <language id="2">
                <![CDATA[]]>
            </language>
        </available_later>
        <associations>
            <categories nodeType="category" api="categories">
                <category xlink:href="http://localhost:8081/develop/PrestaShop/api/categories/10">
                    <id>
                        <![CDATA[10]]>
                    </id>
                </category>
            </categories>
            <images nodeType="image" api="images"/>
            <combinations nodeType="combination" api="combinations"/>
            <product_option_values nodeType="product_option_value" api="product_option_values"/>
            <product_features nodeType="product_feature" api="product_features"/>
            <tags nodeType="tag" api="tags"/>
            <stock_availables nodeType="stock_available" api="stock_availables">
                <stock_available xlink:href="http://localhost:8081/develop/PrestaShop/api/stock_availables/64">
                    <id>
                        <![CDATA[64]]>
                    </id>
                    <id_product_attribute>
                        <![CDATA[0]]>
                    </id_product_attribute>
                </stock_available>
            </stock_availables>
            <attachments nodeType="attachment" api="attachments"/>
            <accessories nodeType="product" api="products"/>
            <product_bundle nodeType="product" api="products"/>
        </associations>
    </product>
</prestashop>
```
But I encoutred an odd behavior :disappointed: even thought the product is linked to its Category, yet in the Product list, the product's category is not displayed, see the attached screen record below:

https://github.com/PrestaShop/PrestaShop/assets/23308427/c037f32e-5610-46a8-b8ff-8c5bca1877cd

Otherwise, the auto tests are :green_circle: https://github.com/hibatallahAouadni/testing_pr/actions/runs/5187558534 (except the develop regression on BO>header)
Please check and feedback.

Thanks!

**Comment by TomLorenzi** [comment]
I may look into it later

**Comment by matks** [comment]
@TomLorenzi Friendly reminder 😉

**Comment by TomLorenzi** [comment]
> @TomLorenzi Friendly reminder 😉

Trying to find some time currrently :(

**Comment by TomLorenzi** [comment]
No time for this sorry, can be closed then.

**Pull request closed without merging**